sympy/ntheory/qs.py

```diff
@@ -0,0 +1,489 @@
+        gamma = factor_base[val].tmem_p * mod_inverse(a // q_l, q_l) % q_l
+        if gamma > q_l / 2:
+            gamma = q_l - gamma
+        B.append(a//q_l*gamma)
+
+    b = sum(B)
+    g = sieve_polynomial([a*a, 2*a*b, b*b - N], a, b)
+
+    for fb in factor_base:
+        if a % fb.prime == 0:
+            continue
+        fb.a_inv = mod_inverse(a, fb.prime)
+        fb.b_ainv = [2*b_elem*fb.a_inv % fb.prime for b_elem in B]
+        fb.soln1 = (fb.a_inv*(fb.tmem_p - b)) % fb.prime
+        fb.soln2 = (fb.a_inv*(-fb.tmem_p - b)) % fb.prime
+    return g, B
+
+def _initialize_ith_poly(N, factor_base, i, g, B):
+    """Initialization stage of ith poly. After we finish sieving 1`st polynomial
+    here we quickly change to the next polynomial from which we will again
+    start sieving. Suppose we generated ith sieve polynomial and now we
+    want to generate (i + 1)th polynomial, where 1 <= i <= 2**(j - 1) - 1
+    where j is the number of prime factors of the coefficient `a`
+    then this function can be used to go to the next polynomial. If
+    i = 2**(j - 1) - 1 then go to _initialize_first_polynomial stage.
+
+    Parameters:
+    ===========
+
+    N : number to be factored
+    factor_base : factor_base primes
+    i : integer denoting ith polynomial
+    g : (i - 1)th polynomial
+    B : array that stores a//q_l*gamma
+    """
+    from sympy import ceiling
+    v = 1
+    j = i
+    while(j % 2 == 0):
+        v += 1
+        j //= 2
+    if ceiling(i / (2**v)) % 2 == 1:
+        neg_pow = -1
+    else:
+        neg_pow = 1
+    b = g.b + 2*neg_pow*B[v - 1]
+    a = g.a
+    g = sieve_polynomial([a*a, 2*a*b, b*b - N], a, b)
+    for fb in factor_base:
+        if a % fb.prime == 0:
+            continue
+        fb.soln1 = (fb.soln1 - neg_pow*fb.b_ainv[v - 1]) % fb.prime
+        fb.soln2 = (fb.soln2 - neg_pow*fb.b_ainv[v - 1]) % fb.prime
+
+    return g
+
+
+def _gen_sieve_array(M, factor_base):
+    """Sieve Stage of the Quadratic Sieve. For every prime in the factor_base
+    that doesn't divide the coefficient `a` we add log_p over the sieve_array
+    such that `-M <= soln1 + i*p <=  M` and `-M <= soln2 + i*p <=  M` where i
+    is an integer. When p = 2 then log_p is only added using
+    `-M <= soln1 + i*p <=  M`.
+
+    Parameters:
+    ===========
+
+    M : sieve interval
+    factor_base : factor_base primes
+    """
+    sieve_array = [0]*(2*M + 1)
+    for factor in factor_base:
+        if factor.soln1 is None: #The prime does not divides a
+            continue
+        for idx in range((M + factor.soln1) % factor.prime, 2*M, factor.prime):
+            sieve_array[idx] += factor.log_p
+        if factor.prime == 2:
+            continue
+        #if prime is 2 then sieve only with soln_1_p
+        for idx in range((M + factor.soln2) % factor.prime, 2*M, factor.prime):
+            sieve_array[idx] += factor.log_p
+    return sieve_array
+
+def _check_smoothness(num, factor_base):
```

Comment:
I think [ntheory's smoothness](https://docs.sympy.org/latest/modules/ntheory.html#sympy.ntheory.factor_.smoothness) function can come handy here.
- Get the smoothness of the number of the number, compare it with the largest factor of the `factor_base` if the former is greater return `false`, else compute powers based on the factor_base.

Reply:
This will be slightly inefficient, the ntheory smoothness will divide with all the primes till the largest factor of the factorbase, which is approx 2* len(factor_base)